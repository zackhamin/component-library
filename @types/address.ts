import { CountryCode } from './countryCode';

export interface Address {
	addressDetails: {
		line1: string;
		line2: string | null;
		city: string;
		county: string | null;
		country: CountryCode;
		postCode: string;
		isDefaultAddress: boolean;
		addressType: string | null;
		addressNickName: string | null;
	};
	addressContactDetails: {
		firstName: string;
		lastName: string;
		mobileNumber: string | null;
	};
	addressMetadata: {
		addressId: string;
		datetimeCreated: Date | string;
		datetimeUpdated: Date | string;
		addressType: string | null;
	};
}
