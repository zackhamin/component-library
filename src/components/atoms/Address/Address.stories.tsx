import { mockAddress } from 'MockData/address';
import { Meta, Story } from '@storybook/react';

import { Address, AddressProps } from './Address';

export default {
	title: 'Atoms/Address',
	component: Address,
} as Meta;

const Template: Story<AddressProps> = (args) => <Address {...args} />;

export const Default = Template.bind({});
Default.args = {
	address: mockAddress,
	options: {
		multipleLines: false,
	},
};

export const DefaultWithNames = Template.bind({});
DefaultWithNames.args = {
	address: mockAddress,
	options: {
		showName: true,
	},
};

export const MultipleLines = Template.bind({});
MultipleLines.args = {
	address: mockAddress,
	options: {
		multipleLines: true,
	},
};

export const MultipleLinesWithNames = Template.bind({});
MultipleLinesWithNames.args = {
	address: mockAddress,
	options: {
		multipleLines: true,
		showName: true,
	},
};

export const WithMobileNumber = Template.bind({});
WithMobileNumber.args = {
	address: mockAddress,
	options: {
		multipleLines: true,
		showMobileNumber: true,
	},
};
