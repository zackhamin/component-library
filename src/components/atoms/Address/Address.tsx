import { Address as AddressType } from '@Types/address';

type ShowMobileNumber = {
	multipleLines: true;
	showMobileNumber?: boolean;
};

type MultipleLines = {
	multipleLines?: false;
	showMobileNumber?: false;
};

type Options = (MultipleLines | ShowMobileNumber) & {
	showName?: boolean;
};

export interface AddressProps {
	address: AddressType;
	options?: Options;
}

export const Address = ({
	address,
	options: { showName, multipleLines, showMobileNumber } = {
		showName: false,
		multipleLines: false,
		showMobileNumber: false,
	},
}: AddressProps) => {
	const {
		addressDetails: { line1, line2, city, county, country, postCode },
		addressContactDetails: { firstName, lastName, mobileNumber },
	} = address;

	if (multipleLines) {
		return (
			<p>
				{showName && (
					<>
						{firstName} {lastName}
						<br />
					</>
				)}
				{line1}
				<br />
				{line2 && (
					<>
						{line2}
						<br />
					</>
				)}
				{city}
				<br />
				{county && (
					<>
						{county}
						<br />
					</>
				)}
				{postCode}
				{country && (
					<>
						<br />
						{country}
					</>
				)}
				{showMobileNumber && (
					<>
						<br />
						<br />
						{mobileNumber}
					</>
				)}
			</p>
		);
	}

	return (
		<p>
			{showName && (
				<>
					{firstName} {lastName},{' '}
				</>
			)}
			{line1}, {line2 && <>{line2}, </>}
			{city}, {county && <>{county}, </>}
			{postCode}
			{country && (
				<>
					, {country}
					<br />
				</>
			)}
		</p>
	);
};
