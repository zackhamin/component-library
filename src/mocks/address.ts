import { Address } from '@Types/address';

export const mockAddress: Address = {
	addressDetails: {
		line1: '11 High Road',
		line2: null,
		city: 'London',
		county: null,
		country: 'GB',
		postCode: 'SW1 2AB',
		isDefaultAddress: false,
		addressType: null,
		addressNickName: null,
	},
	addressContactDetails: {
		firstName: 'John',
		lastName: 'Doe',
		mobileNumber: '07712345678',
	},
	addressMetadata: {
		addressId: '8d063d44-1e27-4a38-9602-552730b90552',
		datetimeCreated: '2022-01-01 00:00:00 UTC',
		datetimeUpdated: '2022-01-01 00:00:00 UTC',
		addressType: null,
	},
};
