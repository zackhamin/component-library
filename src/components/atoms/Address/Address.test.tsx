import { mockAddress } from 'MockData/address';
import { render } from '@testing-library/react';

import { Address } from './Address';

describe('Address', () => {
	it('renders correctly', () => {
		const { container } = render(<Address address={mockAddress} />);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders multi line address correctly', () => {
		const { container } = render(
			<Address address={mockAddress} options={{ multipleLines: true }} />
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders address with name correctly', () => {
		const { container } = render(
			<Address
				address={mockAddress}
				options={{ showName: true, multipleLines: true }}
			/>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders multi line address with mobile number correctly', () => {
		const { container } = render(
			<Address
				address={mockAddress}
				options={{
					showName: true,
					multipleLines: true,
					showMobileNumber: true,
				}}
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});
});
