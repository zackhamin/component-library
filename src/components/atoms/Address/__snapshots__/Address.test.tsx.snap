// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address renders address with name correctly 1`] = `
<p>
  John
   
  Doe
  <br />
  11 High Road
  <br />
  London
  <br />
  SW1 2AB
  <br />
  GB
</p>
`;

exports[`Address renders correctly 1`] = `
<p>
  11 High Road
  , 
  London
  , 
  SW1 2AB
  , 
  GB
  <br />
</p>
`;

exports[`Address renders multi line address correctly 1`] = `
<p>
  11 High Road
  <br />
  London
  <br />
  SW1 2AB
  <br />
  GB
</p>
`;

exports[`Address renders multi line address with mobile number correctly 1`] = `
<p>
  John
   
  Doe
  <br />
  11 High Road
  <br />
  London
  <br />
  SW1 2AB
  <br />
  GB
  <br />
  <br />
  07712345678
</p>
`;
